src/store/safeAppsSlice.ts

```diff
@@ -19,26 +21,50 @@ export const safeAppsSlice = createSlice({
   reducers: {
     setPinned: (state, { payload }: PayloadAction<{ chainId: string; pinned: SafeAppsPerChain['pinned'] }>) => {
       const { pinned, chainId } = payload
-      state[chainId] ??= { pinned: [] }
+
+      // Initialise chain-specific state
+      state[chainId] ??= { pinned: [], opened: [] }
+      // If apps were opened before any were pinned, no pinned array exists
+      state[chainId].pinned ??= []
+
       state[chainId].pinned = pinned
     },
-    setSafeApps: (state, { payload }: PayloadAction<SafeAppsState>) => {
+    markOpened: (state, { payload }: PayloadAction<{ chainId: string; id: SafeAppData['id'] }>) => {
+      const { id, chainId } = payload
+
+      // Initialise chain-specific state
+      state[chainId] ??= { pinned: [], opened: [] }
+      // If apps were pinned before any were opened, no opened array exists
+      state[chainId].opened ??= []
```

Comment:
I extended the `pinned` state which is also per `chainId`.

Reply:
I can see that but it doesn't answer my question.

Reply:
As per our discussion, I've adjusted it to mark open apps across all chains. Sadly this complicates the types and requires some casting but it otherwise works nicely.